File: leetcodeProblems/word_search_2.js
/**
 * @param {character[][]} board
 * @param {string[]} words
 * @return {string[]}
 */

var findWords = function(board, words) {
    let res = [];
  
    const root = buildTrie(words);
    for (let row = 0; row < board.length; row++) {
      for (let col = 0; col < board[0].length; col++) {
        dfs(root, row, col, board, res);
      }
    }
    return res;
  
  };
  
    function dfs(node, row, col, board, res) {
      if (node.end) {
        res.push(node.end);
        node.end = null;   // make sure only print one time for each word
      }
  
      if (row < 0 || row >= board.length || col < 0 || col >= board[0].length) return;
      if (!node[board[row][col]]) return;
  
      const c = board[row][col];
      board[row][col] = '#';  // mark visited
      dfs(node[c], row + 1, col, board, res); // up
      dfs(node[c], row - 1, col, board, res); // down
      dfs(node[c], row, col + 1, board, res); // right
      dfs(node[c], row, col - 1, board, res); // left
      board[row][col] = c;  // reset - back track
    }
  
  function buildTrie(words) {
      const root = {};
      for (let w of words) {
        let pointer = root; // here 'pointer' just a reference, that we use to go down from root till last child node
                            // and when we rich last child node - this is the end of the world
                            // and instead of setting "node.end = true", we set "node.end = word"
        for (let c of w) {
          if (!pointer[c]) pointer[c] = {}; // if we already have such node, lets ignore it creating and just move the pointer
          pointer = pointer[c];
        }
        pointer.end = w; 
      }
      return root;
    }

let board = [
    ['o','a','a','n'],
    ['e','t','a','e'],
    ['i','h','k','r'],
    ['i','f','l','v']
  ]
  
let words = ["oath","pea","eat","rain", "oat", "rainy", "me"]

console.log(findWords(board, words))

File: leetcodeProblems/word_search_2.py
"""
    Problem Statement:
        Given a 2D board and a list of words from the dictionary, find all words in the board.
        Each word must be constructed from letters of sequentially adjacent cell, where "adjacent" cells are
        those horizontally or vertically neighboring. The same letter cell may not be used more than once in a word.

    Trie Data Structure Sample:
        {
            'p': {'e': {'a': {'*': True}}}, 
            'r': {'a': {'i': {'n': {'*': True}}}}, 
            'e': {'a': {'t': {'*': True}}}, 
            'o': {'a': {'t': {'h': {'*': True}, '*': True}}}
        }
"""

class Solution():

    def findWords(self, board, words):
        wordsmatched = set()
        #build the trie datastructure
        trie = self.build_trie(words)
        rows, cols = len(board), len(board[0])

        for row in range(rows):
            for col in range(cols):
                if board[row][col] in trie:
                    check_words = self.search(row, col, board, trie)
                    
                    #find all the words
                    for word in check_words:
                        wordsmatched.add(word)

        return wordsmatched
    
    def build_trie(self, words):
        trie = {}
        for word in words:
            curr = trie
            for char in word:
                curr = curr.setdefault(char, {})

            curr['*'] = True
        
        return trie
    
    def search(self, row, col, board, trie):
        matchedwords = []
        rows, cols = len(board), len(board[0])

        directions = [(-1, 0), (1, 0), (0, -1), (0, 1)] #up, down, left, right
        stack = [(row, col, trie, [], set())]

        while stack:
            curr_row, curr_col, curr_trie, matched_word, seen = stack.pop()
            #create a copy for overwrite
            curr_trie, matched_word, seen = curr_trie.copy(), matched_word.copy(), seen.copy()

            #get the current character
            char = board[curr_row][curr_col]

            if char not in curr_trie:
                continue
            matched_word.append(char)
            seen.add((curr_row, curr_col)) #mark the cell as seen

            if '*' in curr_trie[char]:
                matchedwords.append("".join(matched_word))
            
            #update the trie pointer
            curr_trie = curr_trie[char]

            for row_offset, col_offset in directions:
                #next row and next col
                next_row, next_col = row_offset + curr_row, col_offset + curr_col
                #condition check
                if 0 <= next_row < rows and 0 <= next_col < cols and (next_row, next_col) not in seen:
                    stack.append((next_row, next_col, curr_trie, matched_word, seen))
        
        return matchedwords

board = [
  ['o','a','a','n'],
  ['e','t','a','e'],
  ['i','h','k','r'],
  ['i','f','l','v']
]

words = ["oath","pea","eat","rain", "oat"]

print(Solution().findWords(board, words))